### DISCUSSION TITLE
Visualization of MDAnalysis results in Blender

### DISCUSSION THREAD

**Discussion by PardhavMaradani** [MDAnalysis/mdanalysis#4862] [GSoC Discussions]
Hi. I was introduced to MDAnalysis by a professor of mine who also encouraged me to attend the Streaming workshop that happened last month. I looked around and one of the projects that I found quite interesting is the visualization of MDAnalysis results in Blender that was presented during the [MDAnalysis and Molecular Nodes workshop](https://www.youtube.com/watch?v=3zKBjnRnAMg) using the very impressive [Molecular Nodes](https://bradyajohnston.github.io/MolecularNodes/) extension by @bradyajohnston and was also a [GSoC 2024 project idea](https://github.com/MDAnalysis/mdanalysis/wiki/GSoC-2024-Project-Ideas#project-5-better-interfacing-of-blender-and-mdanalysis). I noticed that there weren't any major updates to this since.

I wanted to take a crack at this now and build a quick proof-of-concept to see what is feasible since a lot has changed in [Blender](https://www.blender.org) and [Molecular Nodes](https://bradyajohnston.github.io/MolecularNodes/) since then and if there is still interest around this idea.

Looking through the [User Guide](https://userguide.mdanalysis.org/stable/index.html), I made a very rough categorization for visualizations as follows:

- [Basic Annotations](#basic-annotations)
- [Advanced Visualizations](#advanced-visualizations)
- [Charting / Plotting and others](#charting--plotting-and-others)

I hacked up some code to visualize these.

## Basic Annotations

This is to annotate any basic MDAnalysis results as visualizations in the universe displayed in Blender. This could include things like atomic measures (atom names, bonds, bond lengths, bond angles, etc), center of mass (of proteins, residues, distances between them, etc), dihedrals, radii of gyration, geometric MDAnalysis selections (`sphzone`, `sphlayer`, etc) as they vary across the trajectory.

These are done similar to how Blender natively displays annotation overlays in the 3D viewport and how popular extensions like `ScreenCast Keys1`, `measureit`, etc display such information using [`blf`](https://docs.blender.org/api/current/blf.html), [`gpu`](https://docs.blender.org/api/current/gpu.html) modules and such.

Here is an example of basic atomic measures:

<img width="1562" alt="mda_bm_atoms-min" src="https://github.com/user-attachments/assets/5ab6bc7c-565c-45ce-8124-4e09a7d1a743" />

Here is a short video of this visualization in Blender:

https://github.com/user-attachments/assets/81b3b420-cc03-4e52-a8d9-38a7db91509b

Here is an example showing the protein center-of-mass, center-of-mass of two residues and the distance between them:

<img width="1562" alt="mda_bm_coms-min" src="https://github.com/user-attachments/assets/52eb9353-af71-4c95-8033-e99884a60d31" />

Here is a short video of this visualization in Blender:

https://github.com/user-attachments/assets/d7a293d9-6e86-480a-814c-c99093e01e50

Here is an example showing the various dihedral angles for a residue:

<img width="1562" alt="mda_bm_dihedrals-min" src="https://github.com/user-attachments/assets/e4410d21-97d9-4ffa-a398-6813556646ac" />

Here is a short video of this visualization in Blender:

https://github.com/user-attachments/assets/072136ba-697e-4def-80dd-f98095331df0

Here is an example showing all the three radii of gyration along with the combined one:

<img width="1562" alt="mda_bm_rogs-min" src="https://github.com/user-attachments/assets/d9a94947-f65b-4a3c-88db-76c580b26d51" />

Here is a short video of this visualization in Blender:

https://github.com/user-attachments/assets/00515eaf-3cdd-4102-ad95-383dba578b01

Here is an example showing the `sphzone` and `sphlayer` selections:

<img width="1562" alt="mda_bm_selections_1-min" src="https://github.com/user-attachments/assets/b4fdda35-5ce7-43cf-8c0d-8150840df64e" />

<img width="1562" alt="mda_bm_selections_2-min" src="https://github.com/user-attachments/assets/bcf0c934-ec3b-4d68-bb5c-765fac0dc76b" />

Here is a short video of this visualization in Blender:

https://github.com/user-attachments/assets/74b4e300-0a55-416f-ac3c-383360867105

All the annotations can be part of the final renders as shown below:

https://github.com/user-attachments/assets/81b4a44a-3104-4701-97e9-1d963b421ae9

## Advanced Visualizations

Blender is an ideal tool to visualize anything in 3D space. Some of the MDAnalysis analysis results that have volumetric outputs like densities, structures, etc could benefit from this.

Here is an example of [pore dimensions with HOLE2](https://userguide.mdanalysis.org/stable/examples/analysis/polymers_and_membranes/hole2.html):

<img width="1562" alt="mda_ha-min" src="https://github.com/user-attachments/assets/76ceaeb9-f795-4755-933c-a2443034241e" />

The data from `.sph` (generated by the `hole` binary) that has the center and radius information for each of the spheres is used directly within Blender to generate the pore surface with the appropriate color coding and the pore center line.

Here is a short video of this visualization in Blender:

https://github.com/user-attachments/assets/9562940f-7c0e-4a03-8468-ad424cfe00f5

Here is an example of [solvent density around a protein](https://userguide.mdanalysis.org/stable/examples/analysis/volumetric/density_analysis.html):

<img width="1562" alt="mda_da_water-min" src="https://github.com/user-attachments/assets/06a07716-4884-47bb-9c3a-175c26ddf2c9" />

The `water.dx` file generated by MDAnalysis is converted to an [OpenVDB](https://www.openvdb.org) file format that Blender natively supports.  This allows interactive visualization of different `iso` levels, slice through the volumetric data (like in 3D MRI data), etc. The example above shows positive `iso` values in blue and negative ones in red.

Here is a short video of this visualization in Blender:

https://github.com/user-attachments/assets/e07d6eca-3834-4dbe-8a6d-6c305dfdd7a7

Here is another example of charge densities using data from [here](https://github.com/BradyAJohnston/MolecularNodes/issues/438):

<img width="1562" alt="mda_da_apbs-min" src="https://github.com/user-attachments/assets/ee3c52a1-4ef3-46ff-b592-570e3311daa5" />

Here is a short video of this visualization in Blender:

https://github.com/user-attachments/assets/e2aac52f-ffed-42b2-a1a1-768fb6e153c8

## Charting / Plotting and others

A lot of charts / plots that are used for MDAnalysis analysis results can also be generated within Blender - especially the ones that are 3D in nature.  For example [distances between residues](https://userguide.mdanalysis.org/stable/examples/analysis/distances_and_contacts/distances_between_selections.html), which is a heatmap that can be visualized in 3D.  This data varying with frames along the trajectory would then be an animated heatmap. I am not sure if there is a lot of value that visualizations of these in Blender would provide apart from some specific use cases.

---

Blender can already be interfaced with Jupyter Notebooks as demonstrated in the [MDAnalysis and Molecular Nodes workshop](https://www.youtube.com/watch?v=3zKBjnRnAMg) and also more recently in the [Blender Conference 2024](https://www.youtube.com/watch?v=umS8jFXpC-o). Blender can be used either in the headless mode or GUI mode as outlined [here](https://kolibril13.github.io/bpy-gallery/n0getting_started/) in this workshop. We cannot expose everything that Blender offers as python API in the headless mode, but we can have some sensible defaults that are usable. A Blender extension (while exposing the python API) could provide a more user-friendly GUI for users who aren't Blender experts and bridge that gap.

Would love to hear what you folks think. If there is interest around this, I could write up a detailed proposal of how this could potentially work.

**Comment by BradyAJohnston** [comment]
I don't have enough internet signal for the next week to watch the videos, but it looks like a seriously impressive amount of work you've already done. There is a lot of potential in leveraging the GPU module for more molecular-specific overlays and I would be very interested in seeing the code behind it.

**Reply by PardhavMaradani** [reply]
Hi Brady, thanks for checking this out. Yes, the Blender's [GPU](https://docs.blender.org/api/current/gpu.html) would be ideal for annotations and overlays because it doesn't create any additional geometry and is pretty light weight. One caveat is that any such overlays have to be rendered to an offscreen buffer and composited with the actual Blender renderings for any outputs, but this can be done programatically. A very basic example using `blf` and `gpu` can be found in Blender's `Text Editor > Templates > Python > Operator Modal Draw` and a more detailed one in the [measureit](https://projects.blender.org/extensions/measureit) extension code that is the basis for my code.

I will share a slightly cleaned up version of the proof-of-concept code for annotations shortly.

**Reply by PardhavMaradani** [reply]
Here is the gist that has code (single file) for the basic annotations shown above:

[https://gist.github.com/PardhavMaradani/13d77d8c107e3d3c17228dc8598b11bd](https://gist.github.com/PardhavMaradani/13d77d8c107e3d3c17228dc8598b11bd)

The gist above has more details about the basic API for now and usage.

All of this can run from Blender's Python Console. Here are the calls to annotate atom info, bond angles, center-of-masses, distance between center-of-masses and canonical dihedrals for the examples shown.  The selection strings are regular selection strings used by MDAnalysis, so this is generic.

```
import bmda as b
b.show_annotations()
b.show_atom_info('u', 'index 1139')
b.show_bond_angle('u', ['index 1143', 'index 1142', 'index 1144'])
b.show_com('u', 'protein')
b.show_com_distance('u', 'resid 1 and name CA', 'resid 129 and name CA')
b.show_canonical_dihedrals('u', 200)
```

Here are the same calls above with a bit of customization that match the ones shown in the above images.

```
import bmda as b
b.show_annotations()
b.show_atom_info('u', 'index 1139', options = {'showRes': True, 'showSeg': True})
b.show_com('u', 'protein', options = {'text': 'Protein|COM'})
b.show_com_distance('u', 'resid 1 and name CA', 'resid 129 and name CA', options = {'text1': 'Res 1|COM', 'text2': 'Res 129|COM'})
```

Font size, font color, line color, text displayed are the only ones that can all be customized for now as shown in the gist.

The universe details are picked up from the `MNSession` attached to the object by Molecular Nodes that makes all of this possible.

**Comment by PardhavMaradani** [comment]
To complete the thread, I made a slightly detailed list of potential visualizations possible based on the examples in the [Analysis](https://userguide.mdanalysis.org/stable/examples/analysis/README.html) section of the [User Guide](https://userguide.mdanalysis.org/stable/index.html).

Any place where `nglview` is currently used is already a candidate. Though `nglview` is a great tool for immediate visualization and basic interactivity within a notebook, it has its own limitations that arise due to WebGL and the browser based nature of it. `Blender`, used in the GUI mode with a notebook, could provide an 'Integrated Visualization Environment' with far better interactivity and control over the display of MDAnalysis results which might help with better insights and validation of statistical analyses. `Blender` can also be used in a headless mode with images / video as outputs into notebook.

## Alignments and RMS fitting

### Aligning structures and trajectories

Since the results of alignment can be made as new universes (either with `mda.Merge` or writing out to files), these can already be loaded using `MolecularNodes` and visualized.

### RMSD and RMSF

RMSD values with respect to refs can be shown as overlays depending on the context as there is no inherent visualization aspect to it. RMSF values as B-Factors can already be visualized using `MolecularNodes`.

## Distances and contacts

### Distances

Distances between atoms, residues, custom selections (between center-of-mass, center-of-geometry, etc) as they vary across a trajectory can be annotated as shown in the PoC above.

### Contacts

Native contacts (with respect to a `refgroup` or `Q1`/`Q2`) and how they vary over a trajectory can be depicted.

## Trajectory similarity

Geometric similarity of trajectories can potentially be visualized by showing individual paths of different trajectories (for atoms, residues, custom selections, etc) either as raw paths or NURBS splines. Since this is a 3ND space, having an option to visualize individual paths might provide insights into similarity/differences evaluated by other methods.

## Structure

### RDF

Since RDF results are volumetric in nature, they can be depicted similar to the ones in the [Volumetric analysis](#volumetric-analyses) section below with varying sphere radii and color ramps.

### Dihedral angle analysis

Dihedral angles (canonical or otherwise) can be annotated as shown in the PoC above which provide a visualization of how they vary over a trajectory.

### Helix analysis

The geometry of helices characterized by `HELNAL` can be depicted as overlays. This includes the local axes, reference axis, global axes and twist, bend, tilt and screw angles.

## Volumetric analyses

Mass, charge and other densities can be depicted as shown in the PoC above for solvent density around a protein. Different iso levels, range of iso levels at different alpha levels, sliced in any direction can help with detailed visual analysis.

## Dimension reduction

Since principal component analysis already includes a `transform` method, the projections of the original trajectory onto any principal component can be visualized as a separate universe.

## Polymers and membranes

The PoC above shows a visualization of pore analysis with HOLE2.

## Hydrogen Bond Analysis

Detected hydrogen bonds by `HydrogenBondAnaysis` (for the whole universe or custom selections) with all the details like donor, acceptor, DA distance and DHA angle can be depicted as they vary over a trajectory.

---

Though I've outlined some of the visualizations above, the goal of any standalone module that could potentially be built is to not implement the above directly, but provide building blocks that serve as a high level interface over the low level Blender API. All of the visualizations above could be built over such building blocks that other MDAnalysis users can also use for their own custom visualizations.

I would love to hear any thoughts / feedback from this community.

**Reply by BradyAJohnston** [reply]
Thanks for the detailed outline. I agree there is a huge amount of potential here, and the approach of small section working (and thus the drawing code figured out) would enable use in all of the other areas much more easily.

The drawing code itself would be very useful in base MN, to enable info to popup with mouse over for particular residues / atoms etc (a commonly requested feature that I knew was impossible without the `gpu` module that I hadn't taken the dive into yet)

**Reply by BradyAJohnston** [reply]
Pinging @yuxuanzhuang to ensure you see this whole thread, as we have started work on a more dedicated notebook integration via ggmolvis: https://github.com/yuxuanzhuang/ggmolvis

**Reply by PardhavMaradani** [reply]
Hi Brady, thanks for the pointer. Hi Yuxuan. I am glad to know that there is interest around this and you folks have already started work. I will take a closer look at this project and see if there is anything I could help with. Thanks

**Comment by yuxuanzhuang** [comment]
That’s an incredible accomplishment, especially with the advanced visualization and density-related aspects.

I’ll take a closer look at the code snippet you shared, but it’s already clear that there’s so much potential.

The main gist behind `ggmolvis` is to create a somewhat unified API and Python classes for visualization. We’re definitely interested in hearing your thoughts, especially given your current work on various analysis types. If you’re interested in learning more about our progress or sharing your insights, feel free to join the Discord channel: https://discord.com/channels/807348386012987462/1256156074008903741.

You’re also welcome to join our next biweekly discussion on the project. Just DM me or Brady on Discord if you’d like to participate.

**Reply by PardhavMaradani** [reply]
Hi Yuxuan, thanks for your response. Yes, I am interested in learning more and will join Discord (please bear with me while I figure things out as this is my first time), look around and reach out to you and Brady.

**Comment by PardhavMaradani** [comment]
Closing this. A lot of items discussed here were implemented as part of the GSoC 2025 [project](https://summerofcode.withgoogle.com/programs/2025/projects/9BR8jbvV). More details about the project, proposal and the contributions can be found in the [final report](https://pardhavmaradani.github.io/posts/gsoc-2025-final-report/) for this GSoC project. Thanks
